revnetgroup/revnetgroup.8
.\" -*- nroff -*-
.TH revnetgroup 8 "March 2001" "YP Server"
.SH NAME
revnetgroup - generate reverse netgroup data
.SH SYNOPSIS
.B @YPBINDIR@/revnetgroup \-u|\-h 
.SH DESCRIPTION
.B revnetgroup
processes the contents of a  netgroup file into a reverse netgroup form.
.LP
.SH OPTIONS
.TP
.B \-u
Generate netgroup.byuser output
.TP
.B \-h
Generate netgroup.byhost outpout
.TP
revnetgroup reads the data from the standard input. 
.SH FILES
/etc/netgroup
.TP
@YPMAPDIR@/Makefile
.SH "SEE ALSO"
.BR makedbm (8),
.BR ypserv (8)
.SH AUTHOR
Thorsten Kukuk <kukuk@suse.de>


NEWS
ypserv NEWS -- history of user-visible changes.

Copyright (C) 1998-2003 Thorsten Kukuk

Please send bug reports, questions and suggestions to <kukuk@suse.de>.

Version 2.8.90
* Fix return code for wrong/invalid map- and domainname
* Make source code more portable

Version 2.8
* Fix compiler warnings
* Fix bug of counting running childs

Version 2.7
* ypserv now forks again for transfering complete NIS maps
* Update to latest autoconf/automake version

Version 2.6
* Fix bug that rpc.ypxfrd doesn't really close maps
* Allow user configureable number of cached database file handles

Version 2.5
* Fix memory leak

Version 2.4
* Minor bug fixes

Version 2.3
* Minor bug fixes

Version 2.2
* Minor bug fixes
* Allow IP numbers as master host
* Lookup master name of NIS map to avoid compare problems
* Fix problems with ypcat

Version 2.1
* Minor bug fixes/typos
* Fix paths in manual pages
* Document removal of tcp_wrapper support
* Fix bug so that ypxfr transfers the whole map in fallback mode
* rpc.yppasswdd can now register itself to a given port

Version 2.0
* Merge bug fixes from Debian

Version 1.99.0
* Rewrite

Version 1.3.10
* Fix various problems with SIGCHLD handling

Version 1.3.9
* Security update

Version 1.3.8
* yppush fix for ndbm support
* yppush fix for Irix 6.5.x NIS slave servers
* rpc.yppasswdd shadow rule fixes
* rpc.yppasswdd fixes for saving/restoring old file modes

Version 1.3.7
* All daemons give away controlling tty.
* Fix possible buffer overflow in rpc.yppasswdd.
* Use yp_msg and not syslog in rpc.yppasswdd for debug messages
* Use strcasecmp in yppush for comparing hostnames.
* Set Endianes for ypxfrd transfer correct.
* Fix file handle corruption through ypproc_all
* Fix error handling for fork() call
* Add support for keys and datum which are longer then 1024 bytes
* ypinit doesn't need a working ypwhich -m
* makedbm: some bug fixes for aliases and handling comments
* Add --debug flag to rpc.yppasswdd
* Add support for merging passwd with shadow file
* Add support for merging group with gshadow
* Minor bug fixes for 64bit platforms
* Bug fixes for libc5
* Add support for changing passwd file by external program

Version 1.3.6
* Fix bug in ypserv.conf parser
* Fix support for gdbm on 64bit platforms
* Fix yp_all return code for empty maps
* Upgrade ypmake to version 0.11 from Olaf Kirch
* Add support for passwd.adjunct
* Allow all characters >= 32

Version 1.3.5
* makedbm bug fixes

Version 1.3.4
* Fix output of ypxfr
* Fix /var/yp/Makefile dependency

Version 1.3.3
* Fix rpc.yppasswdd seg.fault in error case
* Make compiling with libc5 working again

Version 1.3.2
* Upate documentation
* Fix NDBM support

Version 1.3.1
* Bugfixes

Version 1.3.0
* Support for ndbm (SunOS 4.x/5.x) (Not complete)
* Add FQDN support (configure --enable-fqdn). All hostnames will
  go through gethostbyname to determine the offical hostname
* Add YP V1 protocol

Version 1.2.8
* Bugfixes

Version 1.2.7
* Bugfixes

Version 1.2.6
* pwupdate and makedbm man pages
* Remove tryresolve, hacker could stop ypserv with it

Version 1.2.5
* Add patches from SuSE distribution

Version 1.2.4
* yppasswd and rpc.yppasswdd bugfixes

Version 1.2.3
* Fix _dns_gethostbyname problem with Linux/libc5

Version 1.2.2
* rpc.yppasswdd bugfixes
* ypMakefile improvements

Version 1.2.1
* Fix ypserv DB file caching
* A lot of rpc.yppasswdd bugfixes

Version 1.2.0
* Cache the DB file handles


revnetgroup/revnetgroup.c
/* Copyright (c) 1996, 1997, 1998, 1999, 2000, 2001, 2006 Thorsten Kukuk
   Author: Thorsten Kukuk <kukuk@suse.de>

   The YP Server is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   version 2 as published by the Free Software Foundation.

   The YP Server is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public
   License along with the YP Server; see the file COPYING. If
   not, write to the Free Software Foundation, Inc., 51 Franklin Street,
   Suite 500, Boston, MA 02110-1335, USA. */

#define _GNU_SOURCE

#if defined(HAVE_CONFIG_H)
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#if defined(HAVE_GETOPT_H)
#include <getopt.h>
#endif

#include "hash.h"
#include "compat.h"

#define PARSE_FOR_USERS 0
#define PARSE_FOR_HOSTS 1

extern void rev_setnetgrent (const char *);
extern void rev_endnetgrent (void);
extern int rev_getnetgrent (char **, char **, char **);

static int insert_netgroup (hash_t ** liste, const char *key, const char *val);
static void usage (int exit_code);

hash_t *input = NULL;
static hash_t **output = NULL;
static hash_t **empty = NULL;

int
main (int argc, char **argv)
{
#define BUFSIZE 8192
  char *host, *user, *domain, *key;
  char buffer[BUFSIZE + 1];
  int hosts;
  hash_t *work
   ;
#ifdef DEBUG_NETGROUP
  FILE *debug_file = NULL;
#endif

  /* Only -u or -h are allowed, --version will exit the program */
  /* --users = -u, --hosts = -h */
  hosts = -1;

  if (argc < 2)
    usage (1);

  while (1)
    {
      int c;
      int option_index = 0;
      static struct option long_options[] =
      {
	{"version", no_argument, NULL, '\255'},
	{"hosts", no_argument, NULL, 'h'},
	{"users", no_argument, NULL, 'u'},
	{"help", no_argument, NULL, '\254'},
	{"usage", no_argument, NULL, '\254'},
	{NULL, 0, NULL, '\0'}
      };

      c = getopt_long (argc, argv, "uh", long_options, &option_index);
      if (c == EOF)
	break;
      switch (c)
	{
	case 'u':
	  hosts = PARSE_FOR_USERS;
	  break;
	case 'h':
	  hosts = PARSE_FOR_HOSTS;
	  break;
	case '\255':
          printf ("revnetgroup (%s) %s", PACKAGE, VERSION);
	  exit (0);
	case '\254':
	  usage (0);
	  break;
	default:
	  usage (1);
	  break;
	}
    }

#ifdef DEBUG_NETGROUP
  debug_file = fopen ("/etc/netgroup", "rt");
  if (debug_file == NULL)
    {
      perror ("Error: could not open /etc/netgroup:");
      exit (0);
    }
#endif

  /* Put the netgroup names in a list: */

#ifdef DEBUG_NETGROUP
  while (fgets (buffer, BUFSIZE, debug_file))
#else
  while (fgets (buffer, BUFSIZE, stdin))
#endif
    {
      char *val;
      char *cptr;

      if (buffer[0] == '#')
	continue;

      if (strlen (buffer) == 0)
	continue;

      /* Replace first '\n' with '\0' */
      if ((cptr = strchr (buffer, '\n')) != NULL)
	*cptr = '\0';

      while (buffer[strlen (buffer) - 1] == '\\')
	{
	  char *s;
#ifdef DEBUG_NETGROUP
	  s = fgets (&buffer[strlen (buffer) - 1],
		     BUFSIZE - strlen (buffer), debug_file);
#else
	  s = fgets (&buffer[strlen (buffer) - 1],
		     BUFSIZE - strlen (buffer), stdin);
#endif
	  if (s == NULL)
	    continue;
	  if ((cptr = strchr (buffer, '\n')) != NULL)
	    *cptr = '\0';
	}

      val = (char *) (strpbrk (buffer, " \t"));
      if (val == NULL)
	continue;
      key = (char *) &buffer;
      *val = '\0';
      val++;
      insert_netgroup (&input, key, val);
#ifdef DEBUG_NETGROUP
      fprintf (stderr, "KEY: [%s]\n", key);
#endif
    }

#ifdef DEBUG_NETGROUP
  fclose (debug_file);
#endif


#ifdef DEBUG_NETGROUP
  fprintf (stderr, "About to enter while loop...\n");
#endif

  /*
   * Find all members of each netgroup and keep track of which
   * group they belong to.
   */
  empty = hash_malloc ();
  output = hash_malloc ();
  work = input;
  while (work != NULL)
    {
      rev_setnetgrent (work->key);

#ifdef DEBUG_NETGROUP
      fprintf (stderr, "Processing: [%s]\n", work->key);
#endif

      while (rev_getnetgrent (&host, &user, &domain) != 0)
	{
	  static char star[] = "*";
	  char *key = NULL;
	  char *dat = NULL;

	  /* what are we processing for? */
	  if (hosts == PARSE_FOR_HOSTS)
	    dat = host;
	  else
	    dat = user;

	  /* empty fields are wildcard fields = use '-'
	   * to prevent entries
	   */
	  if (dat == NULL)
	    dat = star;

	  /* if we have an entry with data... */
	  if (dat[0] != '-')
	    {
	      /* create the dat/domain key */
	      if (domain == NULL)
		{
		  key = malloc (strlen (dat) + 3);
		  sprintf (key, "%s.*", dat);
		}
	      else
		{
		  key = malloc (strlen (dat) + strlen (domain) + 2);
		  sprintf (key, "%s.%s", dat, domain);
		}

	      /* if we have a wildcard search */
	      if (*dat == '*')
		{
		  char *val = hash_search (empty, key);

		  if (val != NULL)
		    {
		      char *buf = NULL;
		      char *buf2 = NULL;
		      int found = 0;

		      buf2 = malloc (strlen (val) + 2);
		      sprintf (buf2, "%s,", val);
		      buf = strtok (buf2, ",");

		      while ((buf != NULL) && (found == 0))
			{
			  found = (strcmp (buf, work->key) == 0);
			  buf = strtok (NULL, ",");
			}

		      free (buf2);

		      if (!found)
			{
			  buf = malloc (strlen (work->key) + strlen (val) + 2);
			  sprintf (buf, "%s,%s", val, work->key);
			  hash_delkey (empty, key);
			  hash_insert (empty, key, buf);
			  free (buf);
			}
		    }
		  else
		    {
		      hash_insert (empty, key, work->key);
		    }
		}
	      else
		{
		  /* non-wild card search */
		  char *val = hash_search (output, key);

		  if (val != NULL)
		    {
		      char *buf = NULL;
		      char *buf2 = NULL;
		      int found = 0;

		      buf2 = malloc (strlen (val) + 2);
		      sprintf (buf2, "%s,", val);
		      buf = strtok (buf2, ",");

		      while ((buf != NULL) && (found == 0))
			{
			  found = (strcmp (buf, work->key) == 0);
			  buf = strtok (NULL, ",");
			}

		      free (buf2);

		      if (!found)
			{
			  buf = malloc (strlen (work->key) + strlen (val) + 2);
			  sprintf (buf, "%s,%s", val, work->key);
			  hash_delkey (output, key);

			  hash_insert (output, key, buf);
			  free (buf);
			}
		    }
		  else
		    {
		      hash_insert (output, key, work->key);
		    }
		}

	      free (key);
	    }
	}
      work = work->next;
      /* Release resources used by the getnetgrent code. */
      rev_endnetgrent ();
    }

#ifdef DEBUG_NETGROUP
  fprintf (stderr, "About to print results...\n");
#endif

  /* Print the results. */
  work = hash_first (output);
  while (work != NULL)
    {
      printf ("%s\t%s\n", work->key, work->val);
      work = hash_next (output, work->key);
    }

  work = hash_first (empty);
  while (work != NULL)
    {
      printf ("%s\t%s\n", work->key, work->val);
      work = hash_next (empty, work->key);
    }

#if 0
  remove_netgroup (&input);
  remove_netgroup (&output);
  remove_netgroup (&empty);
#endif

  return 0;
}

static void
usage (int exit_code)
{
  fprintf (stderr, "usage: revnetgroup -u|-h\n");
  fprintf (stderr, "       revnetgroup --version\n");
  exit (exit_code);
}

static int
insert_netgroup (hash_t ** liste, const char *key, const char *val)
{
  hash_t *work;

  work = malloc (sizeof (hash_t));
  work->next = *liste;
  work->key = malloc (strlen (key) + 1);
  work->val = malloc (strlen (val) + 1);
  strcpy (work->key, key);
  strcpy (work->val, val);
  *liste = work;

  return 0;
}
